Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2Connection.swift

```diff
@@ -0,0 +1,146 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP2
+
+protocol HTTP2ConnectionDelegate {
+    func http2ConnectionStreamClosed(_: HTTP2Connection, availableStreams: Int)
+    func http2ConnectionClosed(_: HTTP2Connection)
+}
+
+class HTTP2Connection {
+    let channel: Channel
+    let multiplexer: HTTP2StreamMultiplexer
+    let logger: Logger
+
+    /// the connection pool that created the connection
+    let delegate: HTTP2ConnectionDelegate
+
+    enum State {
+        case starting(EventLoopPromise<Void>)
+        case active(HTTP2Settings)
+        case closed
+    }
+
+    var readyToAcceptConnectionsFuture: EventLoopFuture<Void>
+
+    var settings: HTTP2Settings? {
+        self.channel.eventLoop.assertInEventLoop()
+        switch self.state {
+        case .starting:
+            return nil
+        case .active(let settings):
+            return settings
+        case .closed:
+            return nil
+        }
+    }
+
+    private var state: State
+    let id: HTTPConnectionPool.Connection.ID
+
+    init(channel: Channel,
+         connectionID: HTTPConnectionPool.Connection.ID,
+         delegate: HTTP2ConnectionDelegate,
+         logger: Logger) throws {
+        precondition(channel.isActive)
+        channel.eventLoop.preconditionInEventLoop()
+
+        let readyToAcceptConnectionsPromise = channel.eventLoop.makePromise(of: Void.self)
+
+        self.channel = channel
+        self.id = connectionID
+        self.logger = logger
+        self.multiplexer = HTTP2StreamMultiplexer(
+            mode: .client,
+            channel: channel,
+            targetWindowSize: 65535,
+            outboundBufferSizeHighWatermark: 8196,
+            outboundBufferSizeLowWatermark: 4092,
+            inboundStreamInitializer: { (channel) -> EventLoopFuture<Void> in
+                struct HTTP2PushNotsupportedError: Error {}
+                return channel.eventLoop.makeFailedFuture(HTTP2PushNotsupportedError())
+            }
+        )
+        self.delegate = delegate
+        self.state = .starting(readyToAcceptConnectionsPromise)
+        self.readyToAcceptConnectionsFuture = readyToAcceptConnectionsPromise.futureResult
+
+        // 1. Modify channel pipeline and add http2 handlers
```

Comment:
It's not clear to me that we should do this on `init`. I tend to be a bit nervous about `init`s that do a lot of work. Would it be better to move this to a, say, `start` method?